crates/uv-resolver/src/lock.rs

```diff
@@ -596,85 +592,98 @@ pub(crate) struct SourceDist {
             // Or should we return an error?
             ResolvedDist::Installed(_) => todo!(),
-            ResolvedDist::Installable(ref dist) => SourceDist::from_dist(dist),
+            ResolvedDist::Installable(ref dist) => {
+                SourceDist::from_dist(dist, &annotated_dist.hashes)
+            }
         }
     }
 
-    fn from_dist(dist: &Dist) -> Result<Option<SourceDist>, LockError> {
+    fn from_dist(dist: &Dist, hashes: &[HashDigest]) -> Result<Option<SourceDist>, LockError> {
         match *dist {
             Dist::Built(_) => Ok(None),
-            Dist::Source(ref source_dist) => SourceDist::from_source_dist(source_dist).map(Some),
+            Dist::Source(ref source_dist) => {
+                SourceDist::from_source_dist(source_dist, hashes).map(Some)
+            }
         }
     }
 
     fn from_source_dist(
         source_dist: &distribution_types::SourceDist,
+        hashes: &[HashDigest],
     ) -> Result<SourceDist, LockError> {
         match *source_dist {
             distribution_types::SourceDist::Registry(ref reg_dist) => {
-                SourceDist::from_registry_dist(reg_dist)
+                SourceDist::from_registry_dist(reg_dist, hashes)
             }
             distribution_types::SourceDist::DirectUrl(ref direct_dist) => {
-                Ok(SourceDist::from_direct_dist(direct_dist))
+                Ok(SourceDist::from_direct_dist(direct_dist, hashes))
             }
             distribution_types::SourceDist::Git(ref git_dist) => {
-                Ok(SourceDist::from_git_dist(git_dist))
+                Ok(SourceDist::from_git_dist(git_dist, hashes))
             }
             distribution_types::SourceDist::Path(ref path_dist) => {
-                Ok(SourceDist::from_path_dist(path_dist))
+                Ok(SourceDist::from_path_dist(path_dist, hashes))
             }
             distribution_types::SourceDist::Directory(ref directory_dist) => {
-                Ok(SourceDist::from_directory_dist(directory_dist))
+                Ok(SourceDist::from_directory_dist(directory_dist, hashes))
             }
         }
     }
 
-    fn from_registry_dist(reg_dist: &RegistrySourceDist) -> Result<SourceDist, LockError> {
-        // FIXME: Is it guaranteed that there is at least one hash?
-        // If not, we probably need to make this fallible.
+    fn from_registry_dist(
+        reg_dist: &RegistrySourceDist,
+        hashes: &[HashDigest],
+    ) -> Result<SourceDist, LockError> {
         let url = reg_dist
             .file
             .url
             .to_url()
             .map_err(LockError::invalid_file_url)?;
-        let hash = Hash::from(reg_dist.file.hashes[0].clone());
-        Ok(SourceDist {
-            url,
-            hash: Some(hash),
-        })
+        let hash = reg_dist
+            .file
+            .hashes
+            .first()
+            .or(hashes.first())
+            .cloned()
+            .map(Hash::from);
```

Comment:
So... IIUC, `hashes` can be two different things. If the distribution comes from a registry, it's the list of hashes served by the registry. If the distribution comes from elsewhere, it's the hashes we _computed_.